rbd/callbacks.go

```diff
@@ -0,0 +1,63 @@
+package rbd
+
+import (
+	"sync"
+)
+
+// The logic of this file is largely adapted from:
+// https://github.com/golang/go/wiki/cgo#function-variables
+//
+// Also helpful:
+// https://eli.thegreenplace.net/2019/passing-callbacks-and-pointers-to-cgo/
+
+// callbacks provides a tracker for data that is to be passed between Go
+// and C callback functions. The Go callback/object may not be passed
+// by a pointer to C code and so instead integer indexes into an internal
+// map are used.
+// Typically the item being added will either be a callback function or
+// a data structure containing a callback function. It is up to the caller
+// to control and validate what "callbacks" get used.
+type callbacks struct {
+	mutex sync.RWMutex
+	cmap  map[int]interface{}
+}
+
+// add a callback/object to the tracker and return a new index
+// for the object.
+func (cb *callbacks) add(v interface{}) int {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	// this approach assumes that there are typically very few callbacks
+	// in play at once and can just use the length of the map as our
+	// index. But in case of collisions we fall back to simply incrementing
+	// until we find a free key like in the cgo wiki page.
+	// If this code ever becomes a hot path there's surely plenty of room
+	// for optimization in the future :-)
+	index := len(cb.cmap) + 1
+	for {
+		if _, found := cb.cmap[index]; !found {
+			break
+		}
+		index++
+	}
+	cb.cmap[index] = v
+	return index
+}
+
+// remove a callback/object given it's index.
+func (cb *callbacks) remove(index int) {
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+	delete(cb.cmap, index)
+}
+
+// lookup returns a mapped callback/object given an index.
+func (cb *callbacks) lookup(index int) interface{} {
+	cb.mutex.RLock()
+	defer cb.mutex.RUnlock()
+	return cb.cmap[index]
+}
+
+var (
+	cb = &callbacks{cmap: make(map[int]interface{})}
```

Comment:
Not sure if there should be a single global `cb`, or just a helper and have packages use that (split callback maps per functionality like snapshot/watchers/...). I would like a `NewCallbacks()` function, as the exposing of the cmap in the tests looks ugly. There is no need to have a different type of cmap, it is always a `map[int]interface{}`.

Thinking about this more, it would definitely be better to move the callback+shim to its own package. We do need callback support in the rados package too, for #88. Probably there are async CephFS operations too.